extensions/api/data-management/contractnegotiation/src/main/java/org/eclipse/dataspaceconnector/api/datamanagement/contractnegotiation/ContractNegotiationApiController.java

```diff
@@ -63,63 +78,104 @@ public ContractNegotiationApiController(Monitor monitor) {
     @Override
     public ContractNegotiationDto getNegotiation(@PathParam("id") String id) {
         monitor.debug(format("Get contract negotiation with id %s", id));
-        return null;
+
+        return Optional.of(id)
+                .map(service::findbyId)
+                .map(it -> transformerRegistry.transform(it, ContractNegotiationDto.class))
+                .filter(Result::succeeded)
+                .map(Result::getContent)
+                .orElseThrow(() -> new ObjectNotFoundException(ContractDefinition.class, id));
     }
 
     @GET
     @Path("/{id}/state")
     @Override
     public String getNegotiationState(@PathParam("id") String id) {
         monitor.debug(format("Get contract negotiation state with id %s", id));
-        return "some state";
-    }
-
-    @POST
-    @Path("/{id}/cancel")
-    @Override
-    public void cancelNegotiation(@PathParam("id") String id) {
-        // TODO move Negotiation to the CANCELLING/CANCELLED state
-        // TODO Throw IllegalStateException if not possible
-        monitor.debug(format("Attempting to cancel contract negotiation with id %s", id));
-    }
-
-    @POST
-    @Path("/{id}/decline")
-    @Override
-    public void declineNegotiation(@PathParam("id") String id) {
-        // TODO move Negotiation to the DECLINING/DECLINED state
-        // TODO Throw IllegalStateException if not possible
-        monitor.debug(format("Attempting to decline contract negotiation with id %s", id));
+        return Optional.of(id)
+                .map(service::getState)
+                .orElseThrow(() -> new ObjectNotFoundException(ContractDefinition.class, id));
     }
 
     @GET
     @Path("/{id}/agreement")
     @Override
     public ContractAgreementDto getAgreementForNegotiation(@PathParam("id") String negotiationId) {
-        //TODO: fetch agreement for negotiation-id
-        return ContractAgreementDto.Builder.newInstance().negotiationId(negotiationId).build();
+        monitor.debug(format("Get contract agreement of negotiation with id %s", negotiationId));
+
+        return Optional.of(negotiationId)
+                .map(service::getForNegotiation)
+                .map(it -> transformerRegistry.transform(it, ContractAgreementDto.class))
+                .filter(Result::succeeded)
+                .map(Result::getContent)
+                .orElseThrow(() -> new ObjectNotFoundException(ContractDefinition.class, negotiationId));
     }
 
     @POST
     @Override
     public String initiateContractNegotiation(NegotiationInitiateRequestDto initiateDto) {
-
         if (!isValid(initiateDto)) {
             throw new IllegalArgumentException("Negotiation request is invalid");
         }
 
-        return "not-implemente"; //will be the negotiation-id
+        var transformResult = transformerRegistry.transform(initiateDto, ContractOfferRequest.class);
+        if (transformResult.failed()) {
+            throw new IllegalArgumentException("Negotiation request is invalid");
```

Comment:
I would suggest covering this case in `ContractNegotiationApiControllerIntegrationTest`

Reply:
It's not possible, since currently there's no possibility that the transformer returns a failure. it was did like this to respect the `Result` api and avoid possible unexpected exceptions in the future.